File: octave/PnP/rastreamento_pontos.m
function [q] = rastreamento_pontos(q_temp,p1,p2,p3,p4,f)

% garante a ordem em que cada ponto do plano da imagem adquirida pelo sensor é representado

% pkg load statistics deve ser carragada no main

q = zeros(size(q_temp));

q1 = [p1,f]; %pixel
q2 = [p2,f]; %pixel
q3 = [p3,f]; %pixel
q4 = [p4,f]; %pixel

q_temp2 = [ q1;
            q2;
            q3;
            q4];

n = size(q_temp,1);

dist = zeros(n,n);

for i=1:n
for j=1:n

dist(i,j)= pdist([q_temp(i,:);q_temp2(j,:)],"euclidean");

end
end

for i=1:n

[min_v , min_j] = min(dist(i,:));

q(i,:) = q_temp2(min_j,:);

end

File: PnP_py/rastreamento_pontos.py
# -*- coding: utf-8 -*-
import numpy as np

def rastreamento_pontos(q_temp,p1,p2,p3,p4,f):
    
    # garante a ordem em que cada ponto do plano da imagem adquirida pelo sensor é representado
   
    q = np.zeros((np.size(q_temp,0),np.size(q_temp,1)))
    
    q1 = np.concatenate((p1,np.array([f])),0) #pixel
    q2 = np.concatenate((p2,np.array([f])),0) #pixel
    q3 = np.concatenate((p3,np.array([f])),0) #pixel
    q4 = np.concatenate((p4,np.array([f])),0) #pixel
    
    q_temp2 = np.array([[q1],
                        [q2],
                        [q3],
                        [q4]])
    
    n = np.size(q_temp,0)
    
    dist = np.zeros((n,n))
    
    for i in range(n):
        for j in range(n):
    
            dist[i,j]= np.linalg.norm(q_temp[i,:]-q_temp2[j,:]) # distância euclidiana entre pontos

    for i in range(n):
    
        min_j = np.argmin(dist[i,:],0)
    
        q[i,:] = q_temp2[min_j,:]
    
    return q